File: 3249/main.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(countGoodNodes([][]int{
		{0,1},
		{0,2},
		{1,3},
		{1,4},
		{2,5},
		{2,6},
	}))
}

func countGoodNodes(edges [][]int) int {
	n := len(edges) + 1
	g := make([][]int, n)
	for _, edge := range edges {
		x, y := edge[0], edge[1]
		g[x] = append(g[x], y)
		g[y] = append(g[y], x)
	}
	res := 0
	var dfs func(node, parent int) int
	dfs = func(node, parent int) int {
		valid := true
		treeSize := 0
		subTreeSize := 0
		for _, child := range g[node] {
			if child != parent {
				size := dfs(child, node)
				if subTreeSize == 0 {
					subTreeSize = size
				} else if size != subTreeSize {
					valid = false
				}
				treeSize += size
			}
		}
		if valid {
			res++
		}
		return treeSize + 1
	}
	dfs(0, -1)
	return res
}

File: 3249/main.rs
struct Solution {}

impl Solution {
    pub fn count_good_nodes(edges: Vec<Vec<i32>>) -> i32 {
        let n = edges.len() + 1;
        let mut g = vec![vec![]; n];
        for edge in edges {
            let (x, y) = (edge[0] as usize, edge[1] as usize);
            g[x].push(y);
            g[y].push(x);
        }
        let mut res = 0;
        fn dfs(node: usize, parent: isize, g:&Vec<Vec<usize>>, res: &mut i32) -> usize {
            let mut valid = true;
            let mut treeSize = 0;
            let mut subTreeSize = 0;
            for &child in &g[node] {
                if child != parent as usize {
                    let size = dfs(child, node as isize, g, res);
                    if subTreeSize == 0 {
                        subTreeSize = size;
                    } else if size != subTreeSize {
                        valid = false;
                    }
                    treeSize += size;
                }
            }
            if valid {
                *res+=1;
            }
            return treeSize + 1;
        }
        dfs(0, -1, &g, &mut res);
        return res;
    }
}

fn main() {
    println!("{}", Solution::count_good_nodes(vec![
        vec![0,1],
        vec![0,2],
        vec![1,3],
        vec![1,4],
        vec![2,5],
        vec![2,6],
    ]))
}